op-node: Span Batch Derivation

ImTei opened the pull request:
## Contexts

This PR contains the derivation code for [Span Batch](https://github.com/ethereum-optimism/optimism/blob/develop/specs/span-batches.md).

Please refer to the **Implementations - Derivation** and **Hard Fork Activation** sections of the Design Docs and **op-node** section of Implementation Design Docs for details & rationales.

This PR has a dependency on the [**Span Batch Type, Encoding, and Decoding** PR](https://github.com/ethereum-optimism/optimism/pull/7288).

## Changes

- Modifications:
    - Modify **ChannelReader** to decode both **SingularBatch** and **SpanBatch**.
    - Verify if the processing batch’s timestamp matches the expected batch type by **spanBatchTime** at **BatchQueue**.
    - **BatchQueue** splits a **SpanBatch** to multiple **SingularBatch**-es.
    - Leave the rest of the derivation stages as-is by passing **SingularBatches**.
    - ^ Please refer to the `op-node` section of Implementation Design Docs for details.
- Added unit tests.
    - ^ Please refer to the `Derivation` category in the Test List Sheet for each test’s details.

ImTei committed:
Add SpanBatch hardfork activation time into rollup config

Add L2GenesisSpanBatchTImeOffset into deploy config
Activate SpanBatch hardfork for e2e tests by OP_E2E_USE_SPAN_BATCH env

Commit:
Implement Span batch encoding and decoding

Define SpanBatch and related types
Rename BatchV1 to SingularBatch
Add unit test cases

ImTei committed:
Implement span batch derivation

protolambda commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -42,24 +47,52 @@ type BatchQueue struct {
 
 	l1Blocks []eth.L1BlockRef
 
-	// batches in order of when we've first seen them, grouped by L2 timestamp
-	batches map[uint64][]*BatchWithL1InclusionBlock
+	// batches in order of when we've first seen them
+	batches []*BatchWithL1InclusionBlock
```
Is the change to a single batch queue, rather than ones grouped by timestamp, required for span-batch processing?

Some previous context you may not be aware of: A previous version of the code implemented this with the slice of batches, but then changed it to a map where items are grouped by timestamp, to optimize the iteration if there are lots of batches.

I like the simplification back to a slice however, as with span-batches a lot of the batches will be ordered anyway, and iterating the batches does not seem like such a big deal. But happy to hear the other arguments for the `map`.

ImTei replied:
@protolambda Changing to a single batch queue is required to allow overlapping span batches. In the original code, overlapping batches may not be in the queue at `nextTimestamp`, so we have to scan every batches from the past.

Additionally, if I understand correctly, there might be a memory leak in the current code since we don't clean up `remaining` batches from the queue after accepting a batch. This code change also resolves this leak.

protolambda commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -42,24 +47,52 @@ type BatchQueue struct {
+	// batches in order of when we've first seen them
+	batches []*BatchWithL1InclusionBlock
+
+	// nextSpan is cached SingularBatches derived from SpanBatch
+	nextSpan []*SingularBatch
```
I like how the span is expanded into singular batches, and the testing is applied to both the batch types.

protolambda commented on op-node/rollup/derive/batch_queue.go:
```diff
@@ -42,24 +47,52 @@ type BatchQueue struct {
+	return nextBatch
+}
+
+func (bq *BatchQueue) advanceEpoch(nextBatch *SingularBatch) {
```
I would rename this to `advanceEpochMaybe`, since it checks if it has to advance, before it actually advances.

protolambda commented on op-node/rollup/derive/batch_queue_test.go:
```diff
@@ -367,6 +533,387 @@ func TestBatchQueueMissing(t *testing.T) {
+		origin:  l1[0],
+	}
+
+	l2Client := testutils.MockL2Client{}
+	var nilErr error
+	for i, batch := range expectedOutputBatches {
+		if batch != nil {
+			blockRef := singularBatchToBlockRef(t, batch, uint64(i+1))
+			payload := singularBatchToPayload(t, batch, uint64(i+1))
+			l2Client.Mock.On("L2BlockRefByNumber", uint64(i+1)).Times(9999).Return(blockRef, &nilErr)
+			l2Client.Mock.On("PayloadByNumber", uint64(i+1)).Times(9999).Return(&payload, &nilErr)
```
When using a mock you it's nice to assert what happens, i.e. `l2Client.Mock.AssertExpectations(t)` after running through the calls somewhere at the end of the test, to see if the calls were all actually made as expected. You can use `.Maybe()` to make a call optional. I see you use it more like a util here, to set up the client in a supporting form, rather than exact call-by-call assertions.

protolambda commented on op-node/rollup/derive/batch_queue_test.go:
```diff
@@ -367,6 +533,387 @@ func TestBatchQueueMissing(t *testing.T) {
+		if batch != nil {
+			blockRef := singularBatchToBlockRef(t, batch, uint64(i+1))
+			payload := singularBatchToPayload(t, batch, uint64(i+1))
+			l2Client.Mock.On("L2BlockRefByNumber", uint64(i+1)).Times(9999).Return(blockRef, &nilErr)
+			l2Client.Mock.On("PayloadByNumber", uint64(i+1)).Times(9999).Return(&payload, &nilErr)
```
Same here

protolambda commented on op-node/rollup/derive/batches.go:
```diff
@@ -141,3 +167,204 @@ func CheckBatch(cfg *rollup.Config, log log.Logger, l1Blocks []eth.L1BlockRef, l
+		return BatchDrop
+	}
+
+	endEpochNum := batch.GetBlockEpochNum(batch.GetBlockCount() - 1)
+	originChecked := false
```
I think this is safe-ish: it's limited by the `origin_bits` length, and can thus not be infinite. But we might end up waiting for a very large span-batch (in terms of covered L1 origins) if one is crafted maliciously like that. We may need to harden it with additional checks/limits here. We can do so in a follow-up PR however.

protolambda commented on op-node/rollup/derive/batches.go:
```diff
@@ -28,14 +31,37 @@ const (
+			log.Error("failed type assertion to SpanBatch")
+			return BatchDrop
+		}
+		if !cfg.IsSpanBatch(batch.Batch.GetTimestamp()) {
+			log.Warn("received SpanBatch before SpanBatch hard fork")
+			return BatchDrop
+		}
```
I think we may want to add a second check like this based on the L1 block the batch was included in, to check that timestamp as well, to really make sure we do not end up processing span-batches too early by accident.

ImTei replied:
```
The span-batch containing L2 blocks after the hard fork, is included in L1 block before the hard fork timestamp.
```
^ Is this the scenario that could be a problem? can you explain why it's a problem? (we already discussed this in the last weekly sync, sorry for the repeated question 😅)

If it's a problem and we have to check the L1 inclusion time, I think we should change the spec for hard fork activation rule described in [superchain-upgrades](https://github.com/ethereum-optimism/optimism/blob/develop/specs/superchain-upgrades.md#l2-block-timestamp-based-activation). what do you think?

protolambda reviewed (commented):
Looks good, left a few comments about testing and edge-cases to harden against. I will do a second review pass later, since the derivation code is quite critical.

protolambda commented:
Adding a do-not-merge label to give others the chance to review as well.

ImTei committed:
Apply suggestions from code reivews

Rename advanceEpoch() to advanceEpochMaybe()
Set number of calls for L2 client mock and assert them in batch queue tests

protolambda commented:
Opened https://github.com/ethereum-optimism/optimism/pull/7621 to resolve conflicts with `develop`

Follow-ups to implement after this PR:
- \[in response to span-batch origin-bits checking]:
  I think this is safe-ish: it's limited by the origin_bits length, and can thus not be infinite.
  But we might end up waiting for a very large span-batch (in terms of covered L1 origins) if one is crafted maliciously like that.
  We may need to harden it with additional checks/limits here. We can do so in a follow-up PR however.
- The Batch typing improved a little, but we should look if we can clean this up more:
  ideally there are less different types of intermediate batch forms.
- Proto moved the fix from last PR that filters span-batches upon reading from the batch-reader to the channel-in-reader,
  since the batch-reader no longer has L1 inclusion data (due to the way span-batches work).
  We still want to avoid span-batch code-paths before the hardfork activates however,
  so it's worth checking the L1 data before converting a raw span-batch into a full one.
- In the batch-checking function we need to review the hardfork activation conditions, 
  and update the specs as well, so the span-batches activate at the right moment (w.r.t. L1 inclusion and L2 block time)
  This should be added to the span-batch derivation specs, and then referenced as a feature of Canyon,
  in the superchain-upgrades spec document.

protolambda commented:
Closing in favor of #7621
